### PR TITLE
Support SplitAndRetry for GpuRangeExec

### PR DESCRIPTION
**Pull request by firestarman**
closes https://github.com/NVIDIA/spark-rapids/issues/8314

This PR adds the `SplitAndRetry` support for `GpuRangeExec`. 

It adds the retry block to the generating of the sequence on GPU, and also does some refactor by moving the core code into a new iterator called `GpuRangeIterator`.

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -984,6 +984,97 @@ case class GpuFastSampleExec(
...
+                GpuColumnVector.from(tab, Array[DataType](LongType))
+              }
+            }
+          }
+        }
+      }
+      assert(iter.hasNext)
+      closeOnExcept(iter.next()) { batch =>
+        // it should has only one batch
+        assert(iter.isEmpty)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:1042]: This is really confusing because you are doing it differently from everything else that we have done with split and retry. Could you at least document very clearly what is happening both here and in `splitRowsNumberInHalf` so it is clear why there should be one and only one batch, and that is okay because the state that keeps track of this is updated externally.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 2]
```diff
@@ -984,6 +984,97 @@ case class GpuFastSampleExec(
...
+  }
+
+  /**
+   * Reduce the input rows number by half, and it returns a Seq with only one element,
+   * that is the half value.
+   */
+  private def splitRowsNumberInHalf: AutoCloseableLong => Seq[AutoCloseableLong] =
+    (rowsNumber) => {
+      withResource(rowsNumber) { _ =>
+        assert(rowsNumber.value > 0)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:1066]: Could we change this code a bit? Just so we don't loop forever trying to split something that can never be split.

I normally would pick a number much larger than 10, but to avoid changing the test 10 feels okay.
```
if (rowsNumber.value < 10) {
  throw new SplitAndRetryOOM(s"GPU OutOfMemory: the number of rows generated is too small to be split ${rowsNumber.value}!")
}
Seq(AutoCloseableLong(rowsNumber.value / 2))
```

**Review by revans2 — COMMENTED** [review]
Really mostly nits.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 3]
```diff
@@ -984,6 +984,97 @@ case class GpuFastSampleExec(
...
+    } else if (bi > 0) {
+      Long.MaxValue
+    } else {
+      Long.MinValue
+    }
+  }
+
+  private val safePartitionStart = getSafeMargin(partitionStart) // inclusive
+  private val safePartitionEnd = getSafeMargin(partitionEnd) // exclusive, unless start == this
+  private[this] var number: Long = safePartitionStart
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:1007]: I know this was `number` before your change, but can we give this a more descriptive name? `currentSafePartitionStart`?

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: Good suggestion, I changed it to `currentPosition`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 2 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: Good suggestion. Updated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 1 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:1047]: I added more comments. Hope it can help understand better what is happening here.

Yes, it is a little different from the normal way.

Luckily we can do this here, I mean, return a Seq of a single element from the split function. Because the input is NOT a already loaded big batch, but a long value indicating the rows number to be generated. We need multiple elements in the iterator to cache all the rows for the input batch after splitting, but we don't for a single long.

By doing this, we can pull out the single batch directly, and leverage the existing state variables to handle the retry result. If we returns more than one batch by the iterator, we need to maintain the iterator for later loops, leading to more code change. (I think it is not a good idea either to pull out all the batches and concatenate them to return).

**Comment by firestarman** [comment]
build

**Comment by res-life** [comment]
The implement is retrying in each `next` of the `Iterator`. There are will be multiple retryings when the slow GPU memory keeps for a while.

Two implementations:
- Wrap the `next` in a `Iterator`.
- Wrap a `Iterator` and reduce the `maxRowCountPerBatch` in the retry `iterator`.

We'd better to discuss the above 2 solutions.

**Comment by firestarman** [comment]
> The implement is retrying in each next of the Iterator. There are will be multiple retryings when the slow GPU memory keeps for a while.

It is ok. Each task, i think, will have at most one retry at a point in time. And the retry framework in JNI will handle the concurrency of the retries in multiple tasks correctly.

> Wrap the next in a Iterator.
Wrap a Iterator and reduce the maxRowCountPerBatch in the retry iterator.

The Retry requires reentrant functions as input, I have no idea how to make an iterator or the its `next` be reentrant.

**Comment by res-life** [comment]
Below is an alternative. And it's a nit.

Expose a setter for `maxRowCountPerBatch` in original iterator.
Wrap the original iterator, code is like:
```
class iterator(var maxRowCountPerBatch) {
  def hasNext;
  def tryNext(maxRowCountPerBatch): ColumnarBatch
  def updateAfterTryNext = {
    number = xx
    done = xx
  }
}

new RetryIterator(iterator) {
  def hasNext = iterator.hasNext
  def next = withRetry(iterator.maxRowCountPerBatch, retryPolicy) { newMaxRowCountPerBatch =>
    iterator.maxRowCountPerBatch = newMaxRowCountPerBatch
    val batch = iterator.tryNext(newMaxRowCountPerBatch)
    iterator.updateAfterTryNext()
    batch
  }
}
```

**Comment by firestarman** [comment]
> Below is an alternative. And it's a nit.
>....
> 
> new RetryIterator(iterator) {
>   def hasNext = iterator.hasNext
>   def next = withRetry(iterator.maxRowCountPerBatch, retryPolicy) { newMaxRowCountPerBatch =>
>     iterator.maxRowCountPerBatch = newMaxRowCountPerBatch
>     val batch = iterator.tryNext(newMaxRowCountPerBatch)
>     iterator.updateAfterTryNext()
>     batch
>   }
> }
> ```

It looks like each call to the `next` function in this new `RetryIterator` will still trigger a retry. It seems not to fix the issue you pointed out. That is "The implement is retrying in each next of the Iterator. "

**Comment by res-life** [comment]
> The implement is retrying in each next of the Iterator. There are will be multiple retryings when the slow GPU memory keeps for a while.

I mean when the GPU memory is low.  In the `next`, it always first use the `maxRowCountPerBatch`, then use half of `maxRowCountPerBatch`.  And the following `next`s all following this logic.

When the memory is low, we can memory the `maxRowCountPerBatch`. so the following `next` will not require a large memory.

**Comment by firestarman** [comment]
> > It looks like each call to the next function in this new RetryIterator will still trigger a retry. It seems not to fix the issue you pointed out. That is "The implement is retrying in each next of the Iterator. "
> 
> Yes, will trigger the `withRetry`. But if requiring small memory, the retry actually not hapen because of no OOM.

I see. You are just suggesting tune the expected rows number each when you try to generate a batch according to retry OOMs. 
This is a logic change, and I would prefer to add it in by a sparate PR. I am not sure it is good since the rows number will go smaller and smaller after OOMs, and has no chance to go back to the original value even there is enough memory later.

File: tests/src/test/scala/com/nvidia/spark/rapids/RangeRetrySuite.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+package com.nvidia.spark.rapids
+
+import com.nvidia.spark.rapids.Arm.withResource
+import com.nvidia.spark.rapids.jni.RmmSpark
+
+class RangeRetrySuite extends RmmSparkRetrySuiteBase {
+
+  test("Gpu range with split and retry OOM") {
```
Comment (abellina) [inline tests/src/test/scala/com/nvidia/spark/rapids/RangeRetrySuite.scala:24]: I think this test looks good, but I would like to see if you could add a test with other edge cases: 0 rows (start == end), negative step, 1 row, and also no retry as well.

Comment (firestarman) [reply tests/src/test/scala/com/nvidia/spark/rapids/RangeRetrySuite.scala]: done

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by abellina — APPROVED** [review]

**Review by revans2 — APPROVED** [review]
